File: javascript/sync_clawhub_git_2a279f.js
#!/usr/bin/env node
// sync_clawhub_git.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway2:scripts/sync_clawhub_git.py
// auch in: Projects@clawhub:clawhub/Skills/sync_clawhub_git.py
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

/**
 * Bidirektionale ClawHub ↔ Git Synchronisation
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { execSync } = require('child_process');

// Konfiguration
// Resolve paths relative to this repository so the helper works both in the
// hosted workspace and in environments without a /workspace mount.
const __dirnameResolved = path.resolve(__dirname);
const WORKSPACE_ROOT = path.resolve(__dirnameResolved, '..');
const CLAWHUB_DIR = path.join(WORKSPACE_ROOT, "skills");
const GIT_DIR = path.join(WORKSPACE_ROOT, "git", "skills");
const BACKUP_DIR = path.join(WORKSPACE_ROOT, "backups", "sync");
const LOG_FILE = path.join(WORKSPACE_ROOT, "logs", "sync-agent.log");
const IGNORED_NAMES = new Set([".git", ".clawhub", "node_modules", "__pycache__", ".pytest_cache"]);
const RESERVED_SKILL_NAMES = new Set(["github-clones", "skills", "backups", ".restore", "git", "Abstraktionen"]);
const PRESERVED_TARGET_NAMES = IGNORED_NAMES;

// Erstelle Verzeichnisse
function ensureDirExists(dirPath) {
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
    }
}
ensureDirExists(GIT_DIR);
ensureDirExists(BACKUP_DIR);
ensureDirExists(path.dirname(LOG_FILE));

// Logging
function log(message, level = "INFO") {
    const timestamp = new Date().toISOString().slice(0, 19).replace('T', ' ');
    const entry = `[${timestamp}] [${level}] ${message}`;
    console.log(entry);
    fs.appendFileSync(LOG_FILE, entry + '\n');
}

// Validierung
function validateSkill(skillDir) {
    /** Prüft Skill-Struktur - SKILL.md required, scripts/ optional */
    const skillName = path.basename(skillDir);
    if (RESERVED_SKILL_NAMES.has(skillName)) {
        log(`Validation failed: ${skillName} is reserved and must not be synced as a skill`, "ERROR");
        return false;
    }
    const skillMdPath = path.join(skillDir, "SKILL.md");
    if (!fs.existsSync(skillMdPath)) {
        log(`Validation failed: ${skillName} missing SKILL.md`, "ERROR");
        return false;
    }
    return true;
}

function isIgnoredPath(filePath) {
    const parts = filePath.split(path.sep);
    return parts.some(part => IGNORED_NAMES.has(part) || part.endsWith(".pyc"));
}

function isGeneratedDuplicatePath(root, relPath) {
    /** Detect generated duplicate folders such as <skill>/<skill> or scripts/scripts. */
    const parts = relPath.split(path.sep).filter(p => p !== '');
    if (parts.length === 0) return false;
    const rootName = path.basename(root);
    if (parts[0] === rootName) return true;
    for (let i = 1; i < parts.length; i++) {
        if (parts[i] === parts[i - 1]) return true;
    }
    return false;
}

function* iterSyncFiles(root) {
    /** Yield files that should participate in sync comparisons. */
    const walk = function* (currentRoot, relRootParts) {
        const entries = fs.readdirSync(currentRoot, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(currentRoot, entry.name);
            const relPathParts = [...relRootParts, entry.name];
            const relPath = path.join(...relPathParts);

            if (entry.isDirectory()) {
                if (IGNORED_NAMES.has(entry.name) || entry.name.startsWith("__pycache__")) {
                    continue;
                }
                if (isGeneratedDuplicatePath(root, relPath)) {
                    continue;
                }
                if (isIgnoredPath(relPath)) {
                    continue;
                }
                yield* walk(fullPath, relPathParts);
            } else {
                if (IGNORED_NAMES.has(entry.name) || entry.name.endsWith(".pyc")) {
                    continue;
                }
                if (isIgnoredPath(relPath)) {
                    continue;
                }
                if (isGeneratedDuplicatePath(root, relPath)) {
                    continue;
                }
                if (entry.name === "SKILL.md" && relPath !== "SKILL.md") {
                    continue;
                }
                if (!entry.isFile()) {
                    continue;
                }
                yield [fullPath, relPath];
            }
        }
    };
    yield* walk(root, []);
}

function resetSyncTarget(target) {
    /** Remove stale synced content before copying, preserving local-only metadata. */
    ensureDirExists(target);
    const entries = fs.readdirSync(target, { withFileTypes: true });
    for (const entry of entries) {
        if (PRESERVED_TARGET_NAMES.has(entry.name)) {
            continue;
        }
        const fullPath = path.join(target, entry.name);
        if (entry.isDirectory() && !entry.isSymbolicLink()) {
            fs.rmSync(fullPath, { recursive: true });
        } else {
            fs.unlinkSync(fullPath);
        }
    }
}

function copySyncFiles(source, target) {
    resetSyncTarget(target);
    for (const [srcFile, relPath] of iterSyncFiles(source)) {
        const destFile = path.join(target, relPath);
        const destDir = path.dirname(destFile);
        ensureDirExists(destDir);
        fs.copyFileSync(srcFile, destFile);
    }
}

// Backup
function createBackup(source, skillName) {
    /** Erstellt Backup eines Skills */
    const timestamp = new Date().toISOString().slice(0, 19).replace(/[:]/g, '').replace('T', '_');
    const backupPath = path.join(BACKUP_DIR, `${skillName}_${timestamp}`);

    // Backup verzeichnis löschen falls es existiert
    if (fs.existsSync(backupPath)) {
        try {
            fs.rmSync(backupPath, { recursive: true });
            log(`Removed existing backup: ${backupPath}`);
        } catch (e) {
            log(`Failed to remove existing backup ${backupPath}: ${e.message}`, "ERROR");
            return false;
        }
    }

    try {
        fs.cpSync(source, backupPath, { recursive: true });
        log(`Backup created: ${backupPath}`);
        return true;
    } catch (e) {
        log(`Backup failed: ${e.message}`, "ERROR");
        return false;
    }
}

// Hash-Vergleich
function getFileHash(filePath) {
    /** SHA256-Hash einer Datei */
    // Ensure the path points to a regular file.
    try {
        const stat = fs.statSync(filePath);
        if (!stat.isFile()) {
            return "";
        }
        const hasher = crypto.createHash('sha256');
        const stream = fs.createReadStream(filePath);
        return new Promise((resolve, reject) => {
            stream.on('data', (chunk) => hasher.update(chunk));
            stream.on('end', () => resolve(hasher.digest('hex')));
            stream.on('error', reject);
        });
    } catch (e) {
        log(`Failed to hash ${filePath}: ${e.message}`, "ERROR");
        return "";
    }
}

async function getFileHashAsync(filePath) {
    try {
        return await getFileHash(filePath);
    } catch (e) {
        log(`Failed to hash ${filePath}: ${e.message}`, "ERROR");
        return "";
    }
}

// Sync Richtung ClawHub → Git
async function syncToGit(skillName, dryRun = true) {
    /** Synchronisiert ClawHub Skill zu Git */
    const source = path.join(CLAWHUB_DIR, skillName);
    const target = path.join(GIT_DIR, skillName);

    if (!validateSkill(source)) {
        return false;
    }

    // Backup vor Änderungen (nur wenn target existiert)
    if (!dryRun && fs.existsSync(target)) {
        createBackup(target, skillName);
    }

    // Änderungen erkennen
    const changes = [];
    for (const [srcFile, relPath] of iterSyncFiles(source)) {
        const tgtFile = path.join(target, relPath);
        if (!fs.existsSync(tgtFile)) {
            changes.push(`ADD ${relPath}`);
        } else {
            const srcHash = await getFileHashAsync(srcFile);
            const tgtHash = await getFileHashAsync(tgtFile);
            if (srcHash !== tgtHash) {
                changes.push(`UPDATE ${relPath}`);
            }
        }
    }

    // Dry-Run Report
    if (dryRun) {
        log(`DRY-RUN: ${skillName} - ${changes.length} changes`);
        for (const change of changes) {
            log(`  ${change}`);
        }
        return true;
    }

    // Echte Synchronisation
    log(`SYNC: ${skillName} - Applying ${changes.length} changes`);
    copySyncFiles(source, target);
    log(`SYNC: ${skillName} - Complete`);
    return true;
}

// Sync Richtung Git → ClawHub
async function syncToClawhub(skillName, dryRun = true) {
    /** Synchronisiert Git Skill zu ClawHub */
    const source = path.join(GIT_DIR, skillName);
    const target = path.join(CLAWHUB_DIR, skillName);

    if (!validateSkill(source)) {
        return false;
    }

    // Backup vor Änderungen (nur wenn target existiert)
    if (!dryRun && fs.existsSync(target)) {
        createBackup(target, skillName);
    }

    // Änderungen erkennen (gleiche Logik wie oben)
    const changes = [];
    for (const [srcFile, relPath] of iterSyncFiles(source)) {
        const tgtFile = path.join(target, relPath);
        if (!fs.existsSync(tgtFile)) {
            changes.push(`ADD ${relPath}`);
        } else {
            const srcHash = await getFileHashAsync(srcFile);
            const tgtHash = await getFileHashAsync(tgtFile);
            if (srcHash !== tgtHash) {
                changes.push(`UPDATE ${relPath}`);
            }
        }
    }

    // Dry-Run Report
    if (dryRun) {
        log(`DRY-RUN: ${skillName} - ${changes.length} changes`);
        for (const change of changes) {
            log(`  ${change}`);
        }
        return true;
    }

    // Echte Synchronisation
    log(`SYNC: ${skillName} - Applying ${changes.length} changes`);
    copySyncFiles(source, target);
    log(`SYNC: ${skillName} - Complete`);
    return true;
}

// Hauptfunktion
async function main() {
    const args = process.argv.slice(2);
    let skillName, direction, dryRun = false, force = false;

    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--skill':
                skillName = args[++i];
                break;
            case '--direction':
                direction = args[++i];
                break;
            case '--dry-run':
                dryRun = true;
                break;
            case '--force':
                force = true;
                break;
        }
    }

    if (!skillName || !direction) {
        console.error("Missing required arguments: --skill and --direction");
        process.exit(1);
    }

    if (direction !== 'to-git' && direction !== 'to-clawhub') {
        console.error("Invalid direction. Use 'to-git' or 'to-clawhub'");
        process.exit(1);
    }

    log(`Starting sync: ${skillName} (${direction})`);

    let success;
    if (direction === 'to-git') {
        success = await syncToGit(skillName, dryRun);
    } else {
        success = await syncToClawhub(skillName, dryRun);
    }

    if (!success) {
        log("Sync failed", "ERROR");
        process.exit(1);
    }

    log("Sync completed");
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});

File: perl5/sync_clawhub_git_2a279f.pl
#!/usr/bin/env perl
# sync_clawhub_git.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway2:scripts/sync_clawhub_git.py
# auch in: Projects@clawhub:clawhub/Skills/sync_clawhub_git.py
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use File::Path qw(make_path remove_tree);
use File::Copy qw(copy);
use File::Find;
use Digest::SHA qw(sha256_hex);
use Getopt::Long;
use Pod::Usage;
use Time::Piece;

# Konfiguration
# Resolve paths relative to this repository so the helper works both in the
# hosted workspace and in environments without a /workspace mount.
my $script_path = __FILE__;
$script_path =~ s|\\|/|g;  # Normalize path separators on Windows
my @path_parts = split '/', $script_path;
splice @path_parts, -2;  # Remove script name and parent dir
my $WORKSPACE_ROOT = join('/', @path_parts);
$WORKSPACE_ROOT =~ s|/$||;  # Remove trailing slash if present

my $CLAWHUB_DIR = "$WORKSPACE_ROOT/skills";
my $GIT_DIR = "$WORKSPACE_ROOT/git/skills";
my $BACKUP_DIR = "$WORKSPACE_ROOT/backups/sync";
my $LOG_FILE = "$WORKSPACE_ROOT/logs/sync-agent.log";

my %IGNORED_NAMES = map { $_ => 1 } qw(.git .clawhub node_modules __pycache__ .pytest_cache);
my %RESERVED_SKILL_NAMES = map { $_ => 1 } qw(github-clones skills backups .restore git Abstraktionen);
my %PRESERVED_TARGET_NAMES = %IGNORED_NAMES;

# Erstelle Verzeichnisse
make_path($GIT_DIR, { verbose => 0 });
make_path($BACKUP_DIR, { verbose => 0 });
make_path((split '/', $LOG_FILE)[0..$#_-1], { verbose => 0 });

# Logging
sub log_message {
    my ($message, $level) = @_;
    $level //= "INFO";
    my $timestamp = localtime->strftime('%Y-%m-%d %H:%M:%S');
    my $entry = "[$timestamp] [$level] $message\n";
    print $entry;
    open(my $fh, '>>', $LOG_FILE) or warn "Could not open log file: $!";
    print $fh $entry;
    close $fh;
}

# Validierung
sub validate_skill {
    my ($skill_dir) = @_;
    my $skill_name = (split '/', $skill_dir)[-1];
    
    if (exists $RESERVED_SKILL_NAMES{$skill_name}) {
        log_message("Validation failed: $skill_name is reserved and must not be synced as a skill", "ERROR");
        return 0;
    }
    
    if (!-f "$skill_dir/SKILL.md") {
        log_message("Validation failed: $skill_name missing SKILL.md", "ERROR");
        return 0;
    }
    
    return 1;
}

sub is_ignored_path {
    my ($path) = @_;
    my @parts = split '/', $path;
    
    for my $part (@parts) {
        return 1 if exists $IGNORED_NAMES{$part};
        return 1 if $part =~ /\.pyc$/;
    }
    
    return 0;
}

sub is_generated_duplicate_path {
    my ($root, $rel_path) = @_;
    my @parts = split '/', $rel_path;
    
    return 0 unless @parts;
    
    my $root_name = (split '/', $root)[-1];
    return 1 if $parts[0] eq $root_name;
    
    for my $i (1..$#parts) {
        return 1 if $parts[$i] eq $parts[$i-1];
    }
    
    return 0;
}

sub iter_sync_files {
    my ($root) = @_;
    my @files;
    
    find(sub {
        my $current_root = $File::Find::dir;
        my $file = $_;
        
        # Skip ignored directories
        return if -d $file && exists $IGNORED_NAMES{$file};
        return if $file =~ /^__pycache__/;
        
        # Build relative path
        my $full_path = "$current_root/$file";
        $full_path =~ s|^$root/?||;
        $full_path =~ s|^/+||;
        
        # Skip if ignored path
        return if is_ignored_path($full_path);
        
        # Skip if generated duplicate
        return if is_generated_duplicate_path($root, $full_path);
        
        # Special handling for SKILL.md
        if ($file eq 'SKILL.md' && $full_path ne 'SKILL.md') {
            return;
        }
        
        # Only include regular files
        if (-f "$File::Find::dir/$file") {
            push @files, {
                full_path => "$File::Find::dir/$file",
                rel_path => $full_path
            };
        }
    }, $root);
    
    return @files;
}

sub reset_sync_target {
    my ($target) = @_;
    make_path($target, { verbose => 0 });
    
    opendir(my $dh, $target) or return;
    my @children = readdir($dh);
    closedir $dh;
    
    for my $child (@children) {
        next if $child eq '.' || $child eq '..';
        next if exists $PRESERVED_TARGET_NAMES{$child};
        
        my $full_path = "$target/$child";
        if (-d $full_path && !-l $full_path) {
            remove_tree($full_path);
        } elsif (-f $full_path || -l $full_path) {
            unlink $full_path;
        }
    }
}

sub copy_sync_files {
    my ($source, $target) = @_;
    reset_sync_target($target);
    
    my @files = iter_sync_files($source);
    for my $file_info (@files) {
        my $src_file = $file_info->{full_path};
        my $rel_path = $file_info->{rel_path};
        my $dest_file = "$target/$rel_path";
        
        # Create parent directory
        my $parent_dir = $dest_file;
        $parent_dir =~ s|/[^/]+$||;
        make_path($parent_dir, { verbose => 0 });
        
        # Copy file
        copy($src_file, $dest_file) or warn "Failed to copy $src_file to $dest_file: $!";
    }
}

# Backup
sub create_backup {
    my ($source, $skill_name) = @_;
    my $timestamp = localtime->strftime('%Y%m%d_%H%M%S');
    my $backup_path = "$BACKUP_DIR/${skill_name}_$timestamp";
    
    # Backup verzeichnis löschen falls es existiert
    if (-e $backup_path) {
        eval {
            remove_tree($backup_path);
            log_message("Removed existing backup: $backup_path");
        };
        if ($@) {
            log_message("Failed to remove existing backup $backup_path: $@", "ERROR");
            return 0;
        }
    }
    
    eval {
        make_path($backup_path, { verbose => 0 });
        copy_recursive($source, $backup_path);
        log_message("Backup created: $backup_path");
        return 1;
    };
    if ($@) {
        log_message("Backup failed: $@", "ERROR");
        return 0;
    }
}

sub copy_recursive {
    my ($source, $target) = @_;
    
    opendir(my $dh, $source) or die "Cannot open directory $source: $!";
    my @items = readdir($dh);
    closedir $dh;
    
    for my $item (@items) {
        next if $item eq '.' || $item eq '..';
        next if exists $IGNORED_NAMES{$item};
        next if $item =~ /\.pyc$/;
        
        my $src_path = "$source/$item";
        my $tgt_path = "$target/$item";
        
        if (-d $src_path) {
            make_path($tgt_path, { verbose => 0 });
            copy_recursive($src_path, $tgt_path);
        } else {
            copy($src_path, $tgt_path) or die "Cannot copy $src_path to $tgt_path: $!";
        }
    }
}

# Hash-Vergleich
sub get_file_hash {
    my ($file_path) = @_;
    
    # Ensure the path points to a regular file.
    return "" unless -f $file_path;
    
    eval {
        open(my $fh, '<', $file_path) or die "Cannot open $file_path: $!";
        binmode $fh;
        my $hash = sha256_hex(do { local $/; <$fh> });
        close $fh;
        return $hash;
    };
    if ($@) {
        log_message("Failed to hash $file_path: $@", "ERROR");
        return "";
    }
}

# Sync Richtung ClawHub → Git
sub sync_to_git {
    my ($skill_name, $dry_run) = @_;
    $dry_run //= 1;
    
    my $source = "$CLAWHUB_DIR/$skill_name";
    my $target = "$GIT_DIR/$skill_name";
    
    return 0 unless validate_skill($source);
    
    # Backup vor Änderungen (nur wenn target existiert)
    if (!$dry_run && -e $target) {
        create_backup($target, $skill_name);
    }
    
    # Änderungen erkennen
    my @changes;
    my @files = iter_sync_files($source);
    for my $file_info (@files) {
        my $src_file = $file_info->{full_path};
        my $rel_path = $file_info->{rel_path};
        my $tgt_file = "$target/$rel_path";
        
        if (!-e $tgt_file) {
            push @changes, "ADD $rel_path";
        } elsif (get_file_hash($src_file) ne get_file_hash($tgt_file)) {
            push @changes, "UPDATE $rel_path";
        }
    }
    
    # Dry-Run Report
    if ($dry_run) {
        log_message("DRY-RUN: $skill_name - " . scalar(@changes) . " changes");
        for my $change (@changes) {
            log_message("  $change");
        }
        return 1;
    }
    
    # Echte Synchronisation
    log_message("SYNC: $skill_name - Applying " . scalar(@changes) . " changes");
    copy_sync_files($source, $target);
    log_message("SYNC: $skill_name - Complete");
    return 1;
}

# Sync Richtung Git → ClawHub
sub sync_to_clawhub {
    my ($skill_name, $dry_run) = @_;
    $dry_run //= 1;
    
    my $source = "$GIT_DIR/$skill_name";
    my $target = "$CLAWHUB_DIR/$skill_name";
    
    return 0 unless validate_skill($source);
    
    # Backup vor Änderungen (nur wenn target existiert)
    if (!$dry_run && -e $target) {
        create_backup($target, $skill_name);
    }
    
    # Änderungen erkennen (gleiche Logik wie oben)
    my @changes;
    my @files = iter_sync_files($source);
    for my $file_info (@files) {
        my $src_file = $file_info->{full_path};
        my $rel_path = $file_info->{rel_path};
        my $tgt_file = "$target/$rel_path";
        
        if (!-e $tgt_file) {
            push @changes, "ADD $rel_path";
        } elsif (get_file_hash($src_file) ne get_file_hash($tgt_file)) {
            push @changes, "UPDATE $rel_path";
        }
    }
    
    # Dry-Run Report
    if ($dry_run) {
        log_message("DRY-RUN: $skill_name - " . scalar(@changes) . " changes");
        for my $change (@changes) {
            log_message("  $change");
        }
        return 1;
    }
    
    # Echte Synchronisation
    log_message("SYNC: $skill_name - Applying " . scalar(@changes) . " changes");
    copy_sync_files($source, $target);
    log_message("SYNC: $skill_name - Complete");
    return 1;
}

# Hauptfunktion
sub main {
    my $skill;
    my $direction;
    my $dry_run = 0;
    my $force = 0;
    my $help = 0;
    
    GetOptions(
        'skill=s' => \$skill,
        'direction=s' => \$direction,
        'dry-run' => \$dry_run,
        'force' => \$force,
        'help|?' => \$help
    ) or pod2usage(2);
    
    pod2usage(1) if $help;
    pod2usage(2) unless $skill && $direction;
    pod2usage(2) unless $direction eq 'to-git' || $direction eq 'to-clawhub';
    
    log_message("Starting sync: $skill ($direction)");
    
    my $success;
    if ($direction eq 'to-git') {
        $success = sync_to_git($skill, $dry_run);
    } else {
        $success = sync_to_clawhub($skill, $dry_run);
    }
    
    if (!$success) {
        log_message("Sync failed", "ERROR");
        exit 1;
    }
    
    log_message("Sync completed");
}

main() unless caller;

__END__

=head1 NAME

sync_clawhub_git.pl - Bidirektionaler Sync ClawHub ↔ Git

=head1 SYNOPSIS

sync_clawhub_git.pl [options]

 Options:
   --skill         Skill name
   --direction     Direction: to-git or to-clawhub
   --dry-run       Nur Änderungen anzeigen
   --force         Ohne Backup
   --help          Zeige diese Hilfe

=head1 OPTIONS

=over 4

=item B<--skill>

Name des Skills, der synchronisiert werden soll.

=item B<--direction>

Richtung der Synchronisation: entweder 'to-git' oder 'to-clawhub'.

=item B<--dry-run>

Wenn angegeben, werden nur die Änderungen angezeigt, ohne tatsächlich zu synchronisieren.

=item B<--force>

Wenn angegeben, wird kein Backup erstellt.

=back

=head1 DESCRIPTION

Dieses Skript führt eine bidirektionale Synchronisation zwischen ClawHub und Git durch.

=cut
